okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java

```diff
@@ -0,0 +1,376 @@
+  final HandshakerSocketFactory handshakerSocketFactory;
+  TransportTracer.Factory transportTracerFactory = TransportTracer.getDefaultFactory();
+
+  ObjectPool<Executor> transportExecutorPool = DEFAULT_TRANSPORT_EXECUTOR_POOL;
+  ObjectPool<ScheduledExecutorService> scheduledExecutorServicePool =
+      SharedResourcePool.forResource(GrpcUtil.TIMER_SERVICE);
+
+  ServerSocketFactory socketFactory = ServerSocketFactory.getDefault();
+  long keepAliveTimeNanos = KEEPALIVE_TIME_NANOS_DISABLED;
+  long keepAliveTimeoutNanos = DEFAULT_KEEPALIVE_TIMEOUT_NANOS;
+  int flowControlWindow = DEFAULT_FLOW_CONTROL_WINDOW;
+  int maxInboundMetadataSize = GrpcUtil.DEFAULT_MAX_HEADER_LIST_SIZE;
+  int maxInboundMessageSize = GrpcUtil.DEFAULT_MAX_MESSAGE_SIZE;
+
+  private OkHttpServerBuilder(
+      SocketAddress address, HandshakerSocketFactory handshakerSocketFactory) {
+    this.listenAddress = Preconditions.checkNotNull(address, "address");
+    this.handshakerSocketFactory =
+        Preconditions.checkNotNull(handshakerSocketFactory, "handshakerSocketFactory");
+  }
+
+  @Internal
+  @Override
+  protected ServerBuilder<?> delegate() {
+    return serverImplBuilder;
+  }
+
+  @VisibleForTesting
+  OkHttpServerBuilder setTransportTracerFactory(TransportTracer.Factory transportTracerFactory) {
+    this.transportTracerFactory = transportTracerFactory;
+    return this;
+  }
+
+  /**
+   * Override the default executor necessary for internal transport use.
+   *
+   * <p>The channel does not take ownership of the given executor. It is the caller' responsibility
+   * to shutdown the executor when appropriate.
+   */
+  public OkHttpServerBuilder transportExecutor(Executor transportExecutor) {
+    if (transportExecutor == null) {
+      this.transportExecutorPool = DEFAULT_TRANSPORT_EXECUTOR_POOL;
+    } else {
+      this.transportExecutorPool = new FixedObjectPool<>(transportExecutor);
+    }
+    return this;
+  }
+
+  /**
+   * Override the default {@link ServerSocketFactory} used to listen. If the socket factory is not
+   * set or set to null, a default one will be used.
+   */
+  public OkHttpServerBuilder socketFactory(ServerSocketFactory socketFactory) {
+    if (socketFactory == null) {
+      this.socketFactory = ServerSocketFactory.getDefault();
+    } else {
+      this.socketFactory = socketFactory;
+    }
+    return this;
+  }
+
+  /**
+   * Sets a custom keepalive time, the delay time for sending next keepalive ping after a period of
+   * inactivity. An unreasonably small value might be increased, and {@code Long.MAX_VALUE} nano
+   * seconds or a very large value will disable keepalive.
+   */
+  public OkHttpServerBuilder keepAliveTime(long keepAliveTime, TimeUnit timeUnit) {
+    Preconditions.checkArgument(keepAliveTime > 0L, "keepalive time must be positive");
+    keepAliveTimeNanos = timeUnit.toNanos(keepAliveTime);
+    keepAliveTimeNanos = KeepAliveManager.clampKeepAliveTimeInNanos(keepAliveTimeNanos);
```

Comment:
KeepAliveManager comments imply to me that "keep alive pings" are used by clients to keep a connection going - how does the OkHttpServer use this value?

Reply:
[gRFC A8](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md) (ignoring Extending for Basic Health Checking) is probably the best introduction to why we do this stuff. The reasoning for client-side is very similar on server-side, which was covered by [gRFC A9](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md) (which doesn't say too much more). In short, there's two potential values: 1) trigger activity so the _network_ (routers, NATs, etc) knows the connection is still in use and 2) detect if the connection has been broken by the network.